expression-language.md

```diff
@@ -0,0 +1,375 @@
+
+```
+int(hop) < int(ttl) and int(hop) < 1000
+```
+
+Has the same syntactical meaning of:
+
+```
+INT(hop) < INT(ttl) AND INT(hop) < 1000
+```
+
+### 2.1 Expression
+
+The root of the expression is the `expression` rule:
+
+```ebnf
+expression ::= value-identifier | literal | unary-operation | binary-operation | function-invocation | ( "(" expression ")" )
+```
+
+Nested expressions MUST be correctly parenthesized.
+
+### 2.2. Value identifiers and literals
+
+Value identifiers in CESQL have the same restrictions of the [Attribute Naming
+Convention][ce-attribute-naming-convention] from the CloudEvents spec. A value identifier MUST NOT be greater than 20
+characters in length.
+
+```ebnf
+lowercase-char ::= [a-z]
+value-identifier ::= lowercase-char ( lowercase-char | digit )*
+```
+
+CESQL defines 3 different literal kinds: integer numbers, `true` or `false` booleans and `''` or `""` delimited strings.
+
+```ebnf
+digit ::= [0-9]
+number-literal ::= digit+
+
+boolean-literal ::= "true" | "false"
+
+string-literal ::= ( "'" ( [^'] | "\'" )* "'" ) | ( '"' ( [^"] | '\"' )* '"')
+
+literal ::= number-literal | boolean-literal | string-literal
+```
+
+Because string literals can be either `''` or `""` delimited, in the former case, the `"` has to be escaped, while in
+the latter the `"` has to be escaped.
+
+### 2.3. Operators
+
+CESQL defines boolean unary and binary operators, arithmetic unary and binary operators and the `LIKE`, `IN`, `EXISTS`
+operators.
+
+```ebnf
+not-operator = "NOT"
+unary-logical-operator = not-operator
+binary-logic-operator = "AND" | "OR" | "XOR"
+
+unary-numeric-operator = "-"
+binary-comparison-operator = "=" | "!=" | "<>" | ">=" | "<=" | "<" | ">"
+binary-numeric-arithmetic-operator = "+" | "-" | "*" | "/" | "%"
+
+like-operator = "LIKE"
+exists-operator = "EXISTS"
+in-operator = "IN"
+
+unary-operation ::= (unary-numeric-operator | unary-logical-operator) expression
+
+binary-operation ::= expression (binary-comparison-operator | binary-numeric-arithmetic-operator | binary-logic-operator) expression
+
+like-operation ::= expression not-operator? like-operator string-literal
+
+exists-operation ::= exists-operator value-identifier
+
+set-expression ::= "(" expression ("," expression)* ")"
+in-operation ::= expression not-operator? in-operator set-expression
+```
+
+Note: the `set-expression` supports only compile-time set definition, hence the language doesn't include a _set_ type
+that can be manipulated at runtime.
+
+### 2.4. Functions invocation
+
+CESQL supports N arity function invocation:
+
+```ebnf
+char ::= [A-Z] | [a-z]
+parameter ::= expression
+
+function-identifier ::= char ( "_" | char )*
+
+parameter-list ::= parameter ("," parameter)*
+function-invocation ::= function-identifier "(" parameter-list? ")"
+```
+
+## 3. Language semantics
+
+### 3.1. Type System
+
+The type system contains 3 types:
+
+- _String_: Sequence of allowable Unicode characters
+- _Integer_: A signed 32-bit integer
```

Comment:
That should be bigger. Since we allow for arithmetic here, it's pretty easy to get into overflow situations even if CloudEvents metadata is constrained to 32-bit value space.

Reply:
Copy paste from the ce spec integer definition?

Reply:
done, check it out now